Enable Cilium policy enforcement on top of other CNI plugins

aanm opened the pull request:
Some notes:

- The first commit is also in the PR https://github.com/cilium/cilium/pull/6248
- Best if reviewed per commit
- Suggestions to rename `policy-enforcement-interface` are more than welcome
~~- Runtime tests will to be added once the CI is adapted to install Flannel. This adaptation will likely be done in a different PR.~~

Missing parts:

- create k8s job, or something else, to add / remove bpf programs for running contains as the cilium-agent is not able to read from other namespaces once it's running inside a container. that's why some parts related with creating an endpoint based on its PID code are commented.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/6267)
<!-- Reviewable:end -->

aanm committed:
daemon: decoupled API errors from backend errors

Decouples the API errors embedded in the backend to allow the
createEndpoint function to be used by other API handlers besides
PutCreateEndpoint.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
api/v1: add Pid to the EndpointChangeRequest definition

Will allow to create an endpoint based on the container ID and container
ID.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
datapath/loader: add helper to delete bpf programs

Allows the bpf program to be deleted for a given interface, it helps
clean up bpf programs from interfaces that were not created by Cilium.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
k8s/utils: detect infra container based on container labels

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
pkg/workloads: add helper to retrieve Pids of running infra containers

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
add policy enforcement mode on top of an existing CNI plugin

This will allow Cilium to be deployed on top of existing CNI plugins
that provide networking.

For example in Flannel, a CNI plugin, it creates an L2 interface that
connects all veth's containers to it. With this CNI plugin, Cilium will
not create the cilium_host, the cilium_net nor the cilium_vxlan
interfaces as all of the networking is managed by flannel. What Cilium
will do instead is to attach a BPF program to the existing cni0
network interface.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
CNI: read previous result with CNI chaining if available

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
clean bpf programs when running with policy enforcement mode

There will be 2 main paths when a user will deploy Cilium on top of
other CNI plugins.

a) The user will install a CNI plugin and will deploy Cilium on top, for
each it will not set the flag `policy-enforcement-clean-up`. This user
has the intention of running Cilium as long as the cluster lives.

b) The user as a cluster with a CNI plugin already installed. The user
wants to test out Cilium to see the security advantages it could give.
For this time of use case, `policy-enforcement-clean-up` should be set
as true which will leave the cluster as it was before installing Cilium,
meaning all policy enforcement will no longer occur once Cilium is
uninstalled.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
plugins/cni: add installation script for policy enforcement mode

When POLICY_ENFORCEMENT_INTERFACE is set, the cni configuration that
chains flannel with Cilium will be installed and uninstalled.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
examples/kubernetes: add options for policy enforcement mode

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
kubernetes/addons: add Flannel tutorial

Signed-off-by: André Martins <andre@cilium.io>

ianvernon commented on pkg/client/endpoint.go:
```diff
@@ -52,6 +52,13 @@ func (c *Client) EndpointCreate(ep *models.EndpointChangeRequest) error {
 	return Hint(err)
 }
 
+// EndpointCreateID creates a new endpoint without an endpoint ID.
```
Nit: function name confused me since it creates an endpoint without an endpoint ID. Can you change it to `CreateEndpointWithoutID`?

aanm replied:
But it does return an endpoint with an ID, I guess the description is not the best one but the function name is not that bad, what do you think?

ianvernon replied:
>  I guess the description is not the best one but the function name is not that bad, what do you think?

Agreed, feel free to just change the description.

ianvernon commented on daemon/daemon_main.go:
```diff
@@ -812,6 +816,28 @@ func initEnv(cmd *cobra.Command) {
+	// is initialized we need to wait until it is initialized so we can attach
+	// the BPF program to it. If Cilium is running as a Kubernetes DaemonSet,
+	// there is also a script waiting for the interface to be created.
+	if option.Config.IsPolicyEnforcementInterfaceSet() {
```
Can you factor this out into a separate function? The length of the Daemon bootstrapping functions is quite long.

ianvernon commented on daemon/daemon_main.go:
```diff
@@ -870,6 +896,28 @@ func runDaemon() {
 		workloads.IgnoreRunningWorkloads()
 	}
 
+	if option.Config.IsPolicyEnforcementInterfaceSet() {
```
Same here.

ianvernon commented on daemon/endpoint.go:
```diff
@@ -882,3 +885,36 @@ func (h *putEndpointIDLabels) Handle(params PatchEndpointIDLabelsParams) middlew
+			IPV6: ip6.String(),
+		},
+		ContainerID: containerID,
+		ID:          int64(binary.BigEndian.Uint16(ip.To4()[2:])),
```
This isn't a problem right now, but this may conflict with the work done in 
https://github.com/cilium/cilium/pull/6080

ianvernon commented on daemon/daemon_main.go:
```diff
@@ -89,6 +92,14 @@ const (
 	apiTimeout = 60 * time.Second
 )
 
+var (
```
Can you move these to `daemon/cleanup.go`?

ianvernon commented on daemon/daemon_main.go:
```diff
@@ -1045,3 +1067,15 @@ func (d *Daemon) instantiateAPI() *restapi.CiliumAPI {
+
+// Clean cleans up everything created by this package. It closes the returned
+// channel once everything is cleaned up.
+func Clean() <-chan struct{} {
```
Move this to `daemon/cleanup.go` as well.

ianvernon commented on pkg/endpoint/bpf.go:
```diff
@@ -873,6 +873,12 @@ func (e *Endpoint) DeleteMapsLocked() []error {
+// DeleteBPFProgramLocked delete the BPF program associated with the endpoint's
+// veth interface.
+func (e *Endpoint) DeleteBPFProgramLocked() error {
+	return loader.DeleteDatapath(context.TODO(), e.IfName, "ingress")
```
Can you add a comment at Debug level with `e.getLogger()` indicating that we are deleting the endpoint's BPF program?

aanm committed:
test: add flannel integration in the CI

Signed-off-by: André Martins <andre@cilium.io>

brb commented on daemon/daemon_main.go:
```diff
@@ -810,14 +820,51 @@ func initEnv(cmd *cobra.Command) {
+// setHostDeviceWhenReady sets the given ifaceName as the HostDevice. If
+// ifaceName is not found, then it will wait forever until the device is
+// created before it assign it to the hostDevice.
+func setHostDeviceWhenReady(ifaceName string) {
```
What happens when the interface goes down, or when it gets re-created by a CNI plugin? Shouldn't we monitor the interface and re-install the BPF programs?

joestringer replied:
Interesting point, we should at least consider testing an upgrade of flannel while running Cilium and ensuring at least that we're eventually consistent.

aanm replied:
Good point, will do it afterwards

joestringer commented on bpf/bpf_lxc.c:
```diff
@@ -719,7 +719,11 @@ int handle_ingress(struct __sk_buff *skb)
 
 	case bpf_htons(ETH_P_ARP):
 		ep_tail_call(skb, CILIUM_CALL_ARP);
+#ifndef POLICY_ENFORCEMENT_MODE
```
A couple of notes here:
* If we don't handle ARP for this new mode, then the `ep_tail_call()` call above should also be wrapped inside this ifdef. If we still handle ARP, then I don't understand the motivation behind ignoring errors when attempting to call into the ARP program in this mode.
* The naming is a bit confusing from datapath perspective: "If policy mode is not defined then drop". I think it'd be clearer if this were more specific, something like `#ifdef HANDLE_ARP`.

joestringer commented on daemon/daemon_main.go:
```diff
@@ -179,11 +182,12 @@ func daemonMain() {
 		fmt.Println(err)
 		os.Exit(-1)
 	}
+	<-interruptCh
```
As mentioned on the other PR, AFAICT `RootCmd.Execute()` will not be disrupted by something like a SIGINT, and even if it were inside `runDaemon()` it calls `log.Fatalf()` so we still never get to this check.

aanm replied:
That's true, any sugestions?

joestringer replied:
Shift it into the end of `runDaemon()`? If you want to register the handler super early, then maybe `registerSigHandler()` can be called in the `Run()` method of `RootCmd`, then the channel passed into runDaemon?

joestringer commented on bpf/init.sh:
```diff
@@ -141,17 +142,20 @@ function setup_proxy_rules()
-		ip -6 route replace table $PROXY_RT_TABLE ${IP6_LLADDR}/128 dev $HOST_DEV1
-		ip -6 route replace table $PROXY_RT_TABLE default via $IP6_LLADDR dev $HOST_DEV1
-	fi
+    # PRE_EXISTING_DEVICE might not have an IPv6 address
```
Please retain the same indentation as the surrounding code. (Hint: This file uses tabs for indentation)

joestringer commented on bpf/init.sh:
```diff
@@ -141,17 +142,20 @@ function setup_proxy_rules()
-		ip -6 route replace table $PROXY_RT_TABLE default via $IP6_LLADDR dev $HOST_DEV1
-	fi
+    # PRE_EXISTING_DEVICE might not have an IPv6 address
+    if [ -z "${PRE_EXISTING_DEVICE}" ]; then
```
This is a bit of a surprising layer to handle the lack of IPv6 support in a CNI we're enforcing policy on top of. The result will be silent breakdown of L7 enforcement for IPv6 traffic. Can we somehow signal to users that this might be occurring?

aanm replied:
How can you break L7 enforcement for IPv6 traffic if the CNI does not create an IPv6 address at all?

joestringer replied:
I was assuming that the CNI could potentially support IPv6 in future.

The existing code checks whether the `$IP6_HOST` variable is set and configures IPv6 depending on that. This new version assumes that if the policy enforcement interface is set, then IPv6 is not supported. If this is the case, nothing is logged.

I would have expected that in the daemon code, we figure out whether IPv6 support is enabled or not and pass down `$IP6_HOST` as `<nil>` in the flannel case, which shouldn't require this kind of change to the `init.sh` script. Then in the logic where we make this decision in the daemon, we could log at info/debug level saying something like "policy enforcement interface is set, disabling IPv6".

Not a big deal since the current CNI we integrate with doesn't support IPv6, I was just thinking that at some point if we want to integrate with a CNI that does support this, then this hunk will need to be reverted and the logic for whether to configure IPv6 moved into the daemon.

joestringer commented on bpf/init.sh:
```diff
@@ -235,18 +239,23 @@ function encap_fail()
 
-setup_veth_pair $HOST_DEV1 $HOST_DEV2
+if [ -n "${PRE_EXISTING_DEVICE}" ]; then
+    HOST_DEV1="${PRE_EXISTING_DEVICE}"
```
nit: spacing

joestringer commented on pkg/endpoint/connector/add.go:
```diff
@@ -156,3 +157,92 @@ func SetupVethWithNames(lxcIfName, tmpIfName string, mtu int, ep *models.Endpoin
+	// _, ip6, err := ipam.AllocateNext("ipv6")
+	// if err != nil {
+	// 	return nil, fmt.Errorf("unable to allocate IPv6 address: %s", err)
+	// }
```
Commented out code?

joestringer commented on daemon/endpoint.go:
```diff
@@ -125,6 +143,56 @@ func (h *getEndpointID) Handle(params GetEndpointIDParams) middleware.Responder
+			"ID can't be defined")
+	case epTemplate.ContainerID == "" && epTemplate.Pid == 0:
+		return api.New(PostEndpointInvalidCode,
+			"ContainerID and Pid cannot be empty")
```
The error message should ideally express that either `ContainerID` or `PID` must be specified.

What happens if both are specified? What if they conflict?

aanm replied:
TBD

joestringer commented on plugins/cilium-cni/cilium-cni.go:
```diff
@@ -324,6 +325,48 @@ func cmdAdd(args *skel.CmdArgs) error {
+		if err != nil {
+			logger.WithError(err).WithFields(logrus.Fields{
+				logfields.ContainerID: args.ContainerID}).Warn("Unable to derive endpoint")
+			return fmt.Errorf("unable to derive endpoint: %s", err)
```
I assume that if someone misconfigures their system to unintentionally have two CNIs, and they're not running flannel, then this is the path it will hit?

Where do we detect that there is a mismatch between the Cilium configuration (enable this new mode) and multiple CNIs?

aanm replied:
>I assume that if someone misconfigures their system to unintentionally have two CNIs, and they're not running flannel, then this is the path it will hit?

That can never happen because we are the ones that setup the configuration and so far the only configuration available is the `flannel` chained with `cilium`, there should be `04-cilium.conflist`  somewhere in this PR

@joestringer I didn't get the last question

joestringer replied:
Ah, so we generate the CNI config file based upon whether the user configured this option in the Cilium DS? That answers my questions entirely. Thanks.

joestringer commented on daemon/daemon_main.go:
```diff
@@ -496,6 +500,12 @@ func init() {
+		"policy-enforcement-interface", "", "Install a BPF program to allow for policy enforcement in the given network interface. Allows to run Cilium on top of other CNI plugins that provide networking, e.g. flannel, where for flannel, this value should be set with 'cni0'. [EXPERIMENTAL]")
+	viper.BindEnv("policy-enforcement-interface", "CILIUM_POLICY_ENFORCEMENT_INTERFACE")
+	flags.BoolVar(&option.Config.PolicyEnforcementCleanUp,
+		"policy-enforcement-clean-up", false, "When used along the policy-enforcement-interface flag, it cleans up all BPF programs installed when Cilium agent is terminated. (default: false)")
```
This seems overly specific, could we reduce this so that either we're in regular network plumbing mode, or alternatively we're in CNI integration mode where the user must specify the device?

Put another way, what is the use case for a user configuring this policy enforcement cleanup flag to enable it, or to alternatively disable it?

aanm replied:
> I'm a user, I'm using flannel and I want to deploy Cilium to check how it works, can you guarantee that you leave the cluster as it was after I remove Cilium?

Answer: yes, set the `policy-enforcement-clean-up` as `true`

@joestringer clear?

joestringer replied:
And the use case for `policy-enforcement-clean-up=false` is to ensure that when Cilium is terminated or upgraded in this mode, then we don't disable policy enforcement in-between? OK, makes sense.

joestringer commented on daemon/daemon_main.go:
```diff
@@ -823,16 +833,61 @@ func initEnv(cmd *cobra.Command) {
+	if option.Config.IsPolicyEnforcementInterfaceSet() && option.Config.PolicyEnforcementCleanUp {
+		pidfile.HandleCleanup(cleanUPWg, cleanUPSig, func() {
+			d.compilationMutex.Lock()
+			loader.DeleteDatapath(context.Background(), defaults.HostDevice, "egress")
```
Does this mean that we remove Cilium policy enforcement during Cilium restart / upgrade?

aanm replied:
Yes, check https://github.com/cilium/cilium/pull/6267#discussion_r238885829

joestringer commented on pkg/pidfile/cleanup_test.go:
```diff
@@ -0,0 +1,35 @@
+	. "gopkg.in/check.v1"
+)
+
+func (pfs *PidfileTestSuite) TestHandleCleanup(c *C) {
```
I'm not sure I fully understand the concurrent usage of `HandleCleanup()`, it would be good to see some unit tests of that functionality too.

aanm replied:
@joestringer What concurrent usage are you referring too?

joestringer replied:
I was under the impression that multiple callers could queue up cleanup functions via `DeferTerminationCleanupFunction()` with the same WaitGroup/chan, and this could cause some race condition or unexpected behaviour. With the newest version in the tree, I no longer see any concerns with this.

joestringer commented on plugins/cilium-cni/cni-install.sh:
```diff
@@ -37,12 +45,40 @@ if [ -f "${CILIUM_CNI_CONF}" ]; then
 else
 	echo "Installing new ${CILIUM_CNI_CONF}..."
-	cat > ${CNI_CONF_NAME} <<EOF
+    if [ -z "${CILIUM_POLICY_ENFORCEMENT_INTERFACE}" ]; then
```
nit: throughout this file (and other shell scripts) the newly inserted code uses tabs rather than spaces, so the indentation is messed up if you use golang or kernel-style tab widths for shell scripts. (Even here on GitHub you can see the mismatch)

joestringer reviewed (changes requested):
One overall thing, calling this "Policy enforcement mode" seems a bit misleading to me, because it implies that if you're not in this mode, you aren't enforcing policy. This is not the case. I have a few alternative suggestions:
* Network/CNI integration mode
* Non-Routing mode (not so great, because even currently in direct routing mode we don't handle routing)
* Network plumbing mode (need to reverse the logic, though this is my favourite so far)

The `policy-enforcement-interface` seems fine, as that's actually the device we'll be using to enforce policy.

I left a bunch of comments on the earlier parts of the PR, and stopped at the WIP commit. See below for more details. I think that the "endpoint POST" parts of this PR could also potentially be split out for proper review by API folks, which would also serve to reduce the LoC of this PR by about 1K (that's half the PR).

aanm committed:
bpf: modified bpf datapath to work on top of a network bridge device

With the define POLICY_ENFORCEMENT_MODE is set when Cilium is running on
top of another CNI, the bpf datapath change will mostly perform
redirects to the ingress network bridge.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
personal changes to run flannel

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
test: skipping tests that don't work with flannel

Signed-off-by: André Martins <andre@cilium.io>

joestringer commented on bpf/bpf_lxc.c:
```diff
@@ -718,8 +735,12 @@ int handle_ingress(struct __sk_buff *skb)
 		ret = DROP_MISSED_TAIL_CALL;
+#else
+		ep_tail_call(skb, CILIUM_CALL_ARP);
+		ret = 0;
```
I think this is doing the opposite of what we intend?

The existing path should not be touched. Now it's instead breaking all ARP connectivity for the existing case, then furthermore it's marking continuation of a program (ie tail call fallthrough) as an acceptable path. `ep_tail_call(...)` should almost always be followed by a `ret = DROP_MISSED_TAIL_CALL`, because this usually means that the BPF program plumbing is messed up. (Only exception I'm aware of is the new WIP ipvlan work)

joestringer commented on bpf/bpf_netdev.c:
```diff
@@ -485,6 +491,14 @@ int from_netdev(struct __sk_buff *skb)
+            union macaddr mac = HOST_IFINDEX_MAC;
+            return arp_respond(skb, &mac, BPF_F_INGRESS);
+        }
+#endif
```
Please shift this down into the switch case below, so we can get trace notification messages for this traffic. Also, whitespace is messed up.

joestringer replied:
I'm also curious, why do we need to respond to ARP for everything coming from the cni device? Doesn't flannel handle L2/L3 for us in this mode?

joestringer commented on bpf/bpf_netdev.c:
```diff
@@ -449,6 +450,10 @@ static inline int handle_ipv4(struct __sk_buff *skb, __u32 src_identity)
 #endif
 
+#ifdef POLICY_ENFORCEMENT_MODE
+    return redirect(HOST_IFINDEX, BPF_F_INGRESS);
```
I don't see the corresponding IPv6 piece, is that because currently supported CNI to run underneath (ie flannel) don't support IPv6?

aanm replied:
Correct. Since I don't have a way to test it, I didn't change any line in the IPv6 case

joestringer commented on bpf/bpf_lxc.c:
```diff
@@ -642,15 +651,23 @@ static inline int handle_ipv4_from_lxc(struct __sk_buff *skb, __u32 *dstID)
 
 		cilium_dbg_capture(skb, DBG_CAPTURE_DELIVERY, HOST_IFINDEX);
+#ifdef POLICY_ENFORCEMENT_MODE
+		return redirect(HOST_IFINDEX, BPF_F_INGRESS);
```
Can you describe briefly the motivation behind this bit? Is this because usually we rely on our own code for the first level of routing/forwarding, and we typically attach that BPF program on egress, but in this case we want to try to trigger the underlying CNI's logic, which typically is executed when packets ingress into the cni0 device?

aanm replied:
that's my understanding as well.

aanm replied:
change define to `HOST_REDIRECT_TO_INGRESS`

joestringer reviewed (commented):
I think I follow the gist of how this is intended to work from the datapath perspective.

The first point below doesn't seem quite right, please take another look.

Do we mess with the L2 addresses that the stack sees? Normally I think we just handle ARP from the containers, in which case we are replying to state the L2 address of the gateway device. Not sure I follow the intent behind responding to ARP for traffic egressing on the cni0 device.

aanm committed:
start endpoint annotator in daemon

Signed-off-by: André Martins <andre@cilium.io>

eloycoto commented on test/Vagrantfile:
```diff
@@ -91,6 +91,7 @@ Vagrant.configure("2") do |config|
                 sh.args = [
                     "k8s#{i}", "192.168.36.1#{i}", "#{$K8S_VERSION}",
                     "#{$IPv6}", "#{$CONTAINER_RUNTIME}"]
+                sh.env = { "CNI_INTEGRATION" => ENV['CNI_INTEGRATION'] }
```
Why you need this env? I do not think that it's really needed at all.

aanm replied:
Because it's used in `test/provision/k8s_install.sh`

eloycoto replied:
To be consistent then, could you use as a parameter like the other ones?

```
NODE=$1
IP=$2
K8S_VERSION=$3
IPv6=$4
CONTAINER_RUNTIME=$5
```

joestringer commented on bpf/init.sh:
```diff
@@ -376,6 +388,9 @@ fi
 POLICY_MAP="cilium_policy_reserved_${ID_HOST}"
 OPTS="-DFROM_HOST -DFIXED_SRC_SECCTX=${ID_HOST} -DSECLABEL=${ID_HOST} -DPOLICY_MAP=${POLICY_MAP}"
+if [ -n "${PRE_EXISTING_DEVICE}" ]; then
+    OPTS+=" -DPOLICY_ENFORCEMENT_MODE"
```
It'd be good to also compile-test this in builds/CI with another line under `LXC_OPTIONS` in `bpf/Makefile`:

```
-DENABLE_IPV6:-DENABLE_IPV4:-DNETWORK_PLUMBING_MODE
```

aanm replied:
@joestringer what do you mean?

joestringer replied:
When you run `make` (or `make -C bpf`), the makefiles go through various combinations of building the datapath and compile-test them:

https://github.com/cilium/cilium/blob/49fdd6684e8ebeb11f014e6d122ddd588b00c6cc/bpf/Makefile#L57-L69

If you add a new line like this, then it will also test the new mode.

```
-DENABLE_IPV6:-DENABLE_IPV4:-DNETWORK_PLUMBING_MODE
```

joestringer commented on bpf/bpf_lxc.c:
```diff
@@ -587,7 +587,12 @@ static inline int handle_ipv4_from_lxc(struct __sk_buff *skb, __u32 *dstID)
 
 		cilium_dbg_capture(skb, DBG_CAPTURE_DELIVERY, HOST_IFINDEX);
+
+#ifdef POLICY_ENFORCEMENT_MODE
```
I still find this rather misleading, as it implies that policy enforcement is not enabled in the alternative path. However, it's trying to communicate that this mode is only doing policy and not the network plumbing. 

If this were reversed to something like the following, then I think it's easier to understand what's going on in the datapath:

```
#ifdef NETWORK_PLUMBING_MODE
		return redirect(HOST_IFINDEX, 0);
#else
		return redirect(HOST_IFINDEX, BPF_F_INGRESS);
#endif
```

(Similarly for other usage)

aanm replied:
@joestringer I'm open for suggestions to rename the C define, do you have anything in mind?

joestringer replied:
A few ideas:
* `NETWORK_INTEGRATION_MODE`
* `CNI_INTEGRATION_MODE`
* `ROUTING_MODE` (not so great, because even currently in direct routing mode we don't handle routing)
* `NETWORK_PLUMBING_MODE` (need to reverse the logic, though this is my favourite so far)

aanm replied:
change define to `HOST_REDIRECT_TO_INGRESS`

joestringer commented on daemon/daemon_main.go:
```diff
@@ -901,16 +916,61 @@ func initEnv(cmd *cobra.Command) {
+		}
+		time.Sleep(time.Second)
+	}
+	defaults.HostDevice = ifaceName
```
Modifying a value in the defaults package at runtime is very surprising. Shouldn't mutable options be held in the `option` package?

joestringer commented on pkg/endpoint/bpf.go:
```diff
@@ -203,6 +204,9 @@ func (e *Endpoint) writeHeaderfile(prefix string, owner Owner) error {
 	fw.WriteString("#define LB_L3\n")
 	fw.WriteString("#define LB_L4\n")
+	if option.Config.IsPolicyEnforcementInterfaceSet() {
+		fw.WriteString("#define POLICY_ENFORCEMENT_MODE\n")
```
This seems to be duplicated with the following change in init.sh:
```
    OPTS+=" -DPOLICY_ENFORCEMENT_MODE"
```

If you intend for this to be a general node configuration option, then this should be placed into `node_config.h` from `writeNetdevHeader()`, and not configured per endpoint nor in the init script.

aanm replied:
Cool, I wasn't aware of that function. Done

joestringer reviewed (commented):
I made a few minor comments below, mostly cleanups that would help readability of the changes. I note that there's still chunks of commented out code particularly in relation to IPv6 codepaths, but also some other bits.

joestringer commented on daemon/daemon_main.go:
```diff
@@ -901,16 +916,61 @@ func initEnv(cmd *cobra.Command) {
 	}
+
+	if option.Config.PolicyEnforcementInterface == "" {
+		option.Config.PolicyEnforcementInterface = viper.GetString("policy-enforcement-interface")
```
Was this rebased against the config changes? Don't we configure these under `option.Config.Populate()` these days?

aanm commented on Documentation/cmdref/cilium-agent.md:
```diff
@@ -78,6 +78,7 @@ cilium-agent [flags]
       --monitor-queue-size int                      Size of the event queue when reading monitor events (default 32768)
       --mtu int                                     Overwrite auto-detected MTU of underlying network
       --nat46-range string                          IPv6 prefix to map IPv4 addresses to (default "0:0:0:0:0:FFFF::/96")
+      --policy-enforcement-interface string         Installs a BPF program to allow for policy enforcement in the given network interface. Allows to run Cilium on top of other CNI plugins that provide networking, e.g. flannel, where for flannel, this value should be set with 'cni0'. [EXPERIMENTAL]
```
change to `flannel-master-device`

aanm commented on bpf/bpf_lxc.c:
```diff
@@ -739,7 +739,11 @@ int handle_ingress(struct __sk_buff *skb)
 
 	case bpf_htons(ETH_P_ARP):
 		ep_tail_call(skb, CILIUM_CALL_ARP);
+#ifndef POLICY_ENFORCEMENT_MODE
```
Check master with ipvlan integration

aanm commented on Documentation/cmdref/cilium-agent.md:
```diff
@@ -78,6 +78,7 @@ cilium-agent [flags]
       --monitor-queue-size int                      Size of the event queue when reading monitor events (default 32768)
       --mtu int                                     Overwrite auto-detected MTU of underlying network
       --nat46-range string                          IPv6 prefix to map IPv4 addresses to (default "0:0:0:0:0:FFFF::/96")
+      --policy-enforcement-clean-up                 When used along the policy-enforcement-interface flag, it cleans up all BPF programs installed when Cilium agent is terminated.
```
change to `flannel-uninstall-on-exit`

aanm commented on plugins/cilium-cni/cni-install.sh:
```diff
@@ -3,7 +3,15 @@
 
 HOST_PREFIX=${HOST_PREFIX:-/host}
-CNI_CONF_NAME=${CNI_CONF_NAME:-05-cilium.conf}
+if [ -z "${CILIUM_POLICY_ENFORCEMENT_INTERFACE}" ]; then
```
change name of `CILIUM_POLICY_ENFORCEMENT_INTERFACE`

aanm commented on plugins/cilium-cni/04-cilium-cni.conflist:
```diff
@@ -0,0 +1,23 @@
+{
```
change filename to `04-flannel-cilium.conflist`

aanm commented on bpf/bpf_lxc.c:
```diff
@@ -1024,6 +1033,12 @@ ipv4_policy(struct __sk_buff *skb, int ifindex, __u32 src_label, int *forwarding
 
 		skb->cb[CB_IFINDEX] = HOST_IFINDEX;
+
+#ifdef POLICY_ENFORCEMENT_MODE
```
change define to `HOST_REDIRECT_TO_INGRESS`

aanm commented on bpf/bpf_netdev.c:
```diff
@@ -399,7 +400,12 @@ static inline int handle_ipv4(struct __sk_buff *skb, __u32 src_identity)
 		/* Let through packets to the node-ip so they are
 		 * processed by the local ip stack */
 		if (ep->flags & ENDPOINT_F_HOST)
+#ifdef POLICY_ENFORCEMENT_MODE
```
change define to `HOST_REDIRECT_TO_INGRESS`

aanm commented on bpf/bpf_netdev.c:
```diff
@@ -424,6 +430,10 @@ static inline int handle_ipv4(struct __sk_buff *skb, __u32 src_identity)
 	}
 #endif
 
+#ifdef POLICY_ENFORCEMENT_MODE
```
change define to `HOST_REDIRECT_TO_INGRESS`

aanm commented on bpf/bpf_netdev.c:
```diff
@@ -489,6 +500,14 @@ int from_netdev(struct __sk_buff *skb)
 #ifdef FROM_HOST
 	if (1) {
+
+#ifdef POLICY_ENFORCEMENT_MODE
```
Check why this is needed. `ENABLE_HOST_MAC_ARP_RESPOND`

aanm commented on contrib/vagrant/start.sh:
```diff
@@ -42,7 +42,7 @@ export 'VM_BASENAME'=${K8S+"k8s"}
 export 'VAGRANT_DEFAULT_PROVIDER'=${VAGRANT_DEFAULT_PROVIDER:-"virtualbox"}
 # Sets the default cilium TUNNEL_MODE to "vxlan"
-export 'TUNNEL_MODE_STRING'=${TUNNEL_MODE_STRING:-"-t vxlan"}
+export 'TUNNEL_MODE_STRING'=${TUNNEL_MODE_STRING:-"-t disabled"}
```
Add this patch to the contrib guide.

aanm commented on daemon/daemon_main.go:
```diff
@@ -38,6 +38,7 @@ import (
 	"github.com/cilium/cilium/pkg/datapath/loader"
 	"github.com/cilium/cilium/pkg/datapath/maps"
 	"github.com/cilium/cilium/pkg/defaults"
+	"github.com/cilium/cilium/pkg/endpoint"
```
Make it as new PR
